### PR TITLE
Push limit into aggregation for DISTINCT ... LIMIT queries

### LINKED ISSUE #7781: Select DISTINCT with LIMIT 10 is doing a full-scan of the database
### Describe the bug

So I noticed that when executing the query `SELECT DISTINCT my_column FROM dataset LIMIT 10` (trough the Python connector), it was taking lots of time and consuming most of my RAM.
I then looked at the query plan, and it seems like its actually doing `GROUP BY my_column` which causes a full-scan, what makes it even worse, is that all 10 values returned are present in the **first** parquet file in the dataset (`pyarrow.Dataset.files[0]`), so it could've just stopped scanning after the first file immediately.

### To Reproduce

```py
import datafusion
import time

start = time.perf_counter()

ctx = datafusion.SessionContext()
ctx.register_dataset('dataset', dataset)
out = ctx.sql("SELECT DISTINCT my_column FROM dataset LIMIT 10")

print('datafusion', time.perf_counter() - start)

print(repr(out.logical_plan()))
print('-' * 10)
print(repr(out.optimized_logical_plan()))
print('-' * 10)
print(repr(out.execution_plan()))
print('-' * 10)
print(repr(out.explain()))
print('-' * 10)
```
```
datafusion 0.0003965760000141927
Limit: skip=0, fetch=10
  Distinct:
    Projection: dataset.canonical_url
      TableScan: dataset
----------
Limit: skip=0, fetch=10
  Aggregate: groupBy=[[dataset.canonical_url]], aggr=[[]]
    TableScan: dataset projection=[canonical_url]
----------
GlobalLimitExec: skip=0, fetch=10
  CoalescePartitionsExec
    LocalLimitExec: fetch=10
      AggregateExec: mode=FinalPartitioned, gby=[canonical_url@0 as canonical_url], aggr=[]
        CoalesceBatchesExec: target_batch_size=8192
          RepartitionExec: partitioning=Hash([canonical_url@0], 8), input_partitions=568
            AggregateExec: mode=Partial, gby=[canonical_url@0 as canonical_url], aggr=[]
              DatasetExec: number_of_fragments=568, projection=[canonical_url]

----------
DataFrame()
+---------------+---------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                        |
+---------------+---------------------------------------------------------------------------------------------+
| logical_plan  | Limit: skip=0, fetch=10                                                                     |
|               |   Aggregate: groupBy=[[dataset.canonical_url]], aggr=[[]]                                   |
|               |     TableScan: dataset projection=[canonical_url]                                           |
| physical_plan | GlobalLimitExec: skip=0, fetch=10                                                           |
|               |   CoalescePartitionsExec                                                                    |
|               |     LocalLimitExec: fetch=10                                                                |
|               |       AggregateExec: mode=FinalPartitioned, gby=[canonical_url@0 as canonical_url], aggr=[] |
|               |         CoalesceBatchesExec: target_batch_size=8192                                         |
|               |           RepartitionExec: partitioning=Hash([canonical_url@0], 8), input_partitions=568    |
|               |             AggregateExec: mode=Partial, gby=[canonical_url@0 as canonical_url], aggr=[]    |
|               |               DatasetExec: number_of_fragments=568, projection=[canonical_url]              |
|               |                                                                                             |
+---------------+---------------------------------------------------------------------------------------------+
None
----------
```

### Expected behavior

It should stop scanning at the first 10 unique values found.

### Additional context

For context, I'm querying a Parquet dataset, and the files are stored locally:

```py
import pyarrow.dataset as ds
dataset = ds.dataset(...)
```

### PR DESCRIPTION
**Pull request by msirek**
## Which issue does this PR close?

Closes #7781.

## Rationale for this change

Evaluation of queries form
  SELECT DISTINCT _column_list_ FROM _table_ LIMIT _n_; 
may read more rows than necessary when performing a grouped hash aggregation.

If a given batch of input rows is seen which contains more group values than
the LIMIT value, switching the aggregation to output mode early allows the
the limit to be reached more quickly and minimizes the number of rows which
need to be processed by the aggregation, or the input stream.

#### Push limit into AggregateExec for DISTINCT with GROUP BY

This commit adds physical plan rewrite rule `LimitedDistinctAggregation`,
but does not wire it up for use by the optimizer. The rule matches a
`LocalLimitExec` or `GlobalLimitExec` operation as the parent of an 
`AggregateExec` which has a group-by, but no aggregate expressions, order-by 
or ordering requirements, or filtering, and pushes the limit into the                     
the `AggregateExec` as a limit hint.     

As the aggregation may be applied in a series of `AggregateExec` 
operations, the limit is also pushed down a chain of direct    
`AggregateExec` decendents having identical grouping columns.

The rule must be applied before distribution requirements are enforced
as that rule may inject other operations in between the different 
`AggregateExec`s. Applying the rule early means only directly-connected  
`AggregateExec`s need to be examined.

The key point of this rule is that it is only legal for cases where not
all rows in the group need to be processed to ensure correctness.

Unit tests for LimitedDistinctAggregation are included.

#### Soft limit for GroupedHashAggregateStream with no aggregate expressions

This commit wires up the LimitedDistinctAggregation rule in the physical
plan optimizer and updates the GroupedHashAggregateStream with an
optional soft limit on the number of `group_values` in a batch. If the
number of `group_values` in a single batch exceeds the limit, the operation
immediately signals the input is done, switches to output mode and emits all groups.

This commit includes sqllogictests for DISTINCT queries with a LIMIT.
    
The CombinePartialFinalAggregate rule is also updated to convey the
limit on the final aggregation to the combined aggregation.

#### Add datafusion.optimizer.enable_distinct_aggregation_soft_limit setting

This commit adds the datafusion.optimizer.enable_distinct_aggregation_soft_limit
configuration setting, which defaults to true. When true, the
LimitedDistinctAggregation physical plan rewrite rule is enabled, which
pushes a LIMIT into a grouped aggregation with no aggregate expressions,
as a soft limit, to emit all grouped values seen so far once the limit is reached.

#### Fix result checking in topk_aggregate benchmark

This commit fixes the logic which validates the rows returned by the benchmark query.
The test was expecting hexadecimal digits in lowercase, but results are uppercase.

#### Make the topk_aggregate benchmark's make_data function public

This commit moves the make_data function, which generates either random or ascending
time series data, to the data_utils module, so it could be shared by other benchmarks.

#### Add benchmark for DISTINCT queries

This commit adds a benchmark for queries using DISTINCT or GROUP BY with a LIMIT clause
and no aggregate expressions. It is intended to test the performance of the 
`LimitedDistinctAggregation` rewrite rule and new limit hint in `GroupedHashAggregateStream`.

## Are these changes tested?

- [x] unit tests for LimitedDistinctAggregation
- [x] unit test for propagating the AggregateExec limit to the result of `CombinePartialFinalAggregate`
- [x] unit test for datafusion.optimizer.enable_distinct_aggregation_soft_limit setting
- [x] sqllogictests to check correct results of AggregateExec with a limit
- [x] criterion benchmarks for DISTINCT queries

## Are there any user-facing changes?

No

## Notes

This is opened as a draft PR. 
PRs for the individual commits can be opened separately if this is too large to review in one PR.

### CODE REVIEW

**Comment by msirek** [comment]
### Results from criterion_benchmark_limited_distinct
#### baseline
custom-measurement-time/distinct_group_by_u64_narrow_limit_10
                        time:   [120.79 ms 121.22 ms 121.67 ms]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
custom-measurement-time/distinct_group_by_u64_narrow_limit_100
                        time:   [124.01 ms 124.55 ms 125.12 ms]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
custom-measurement-time/distinct_group_by_u64_narrow_limit_1000
                        time:   [124.04 ms 124.48 ms 124.94 ms]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
custom-measurement-time/distinct_group_by_u64_narrow_limit_10000
                        time:   [127.71 ms 129.49 ms 131.52 ms]
Found 15 outliers among 100 measurements (15.00%)
  9 (9.00%) high mild
  6 (6.00%) high severe
Benchmarking custom-measurement-time/group_by_multiple_columns_limit_10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 40.0s. You may wish to increase target time to 136.5s, or reduce sample count to 20.
custom-measurement-time/group_by_multiple_columns_limit_10
                        time:   [1.3416 s 1.3465 s 1.3517 s]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild


#### new

custom-measurement-time/distinct_group_by_u64_narrow_limit_10
                        time:   [83.726 ms 84.168 ms 84.596 ms]
                        **change: [-31.003% -30.568% -30.128%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) low mild
  1 (1.00%) high mild
custom-measurement-time/distinct_group_by_u64_narrow_limit_100
                        time:   [112.00 ms 112.33 ms 112.66 ms]
                        **change: [-10.299% -9.8102% -9.3399%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild
custom-measurement-time/distinct_group_by_u64_narrow_limit_1000
                        time:   [112.01 ms 112.37 ms 112.73 ms]
                        **change: [-10.152% -9.7257% -9.2833%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild
custom-measurement-time/distinct_group_by_u64_narrow_limit_10000
                        time:   [111.86 ms 112.21 ms 112.55 ms]
                        **change: [-14.699% -13.350% -12.114%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
Benchmarking custom-measurement-time/aggregate_group_by_multiple_columns_limit_10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 40.0s. You may wish to increase target time to 114.8s, or reduce sample count to 30.
custom-measurement-time/group_by_multiple_columns_limit_10
                        time:   [1.1212 s 1.1247 s 1.1285 s]
                        **change: [-16.881% -16.473% -16.075%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

### Results from criterion_benchmark_limited_distinct_sampled
#### baseline
Benchmarking distinct query with 100 partitions and 100000 samples per partition with limit 10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 15.3s, or reduce sample count to 30.
distinct query with 100 partitions and 100000 samples per partition with limit 10
                        time:   [151.67 ms 151.93 ms 152.19 ms]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) low mild

Benchmarking distinct query with 10 partitions and 1000000 samples per partition with limit 10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 97.2s, or reduce sample count to 10.
distinct query with 10 partitions and 1000000 samples per partition with limit 10
                        time:   [922.72 ms 933.79 ms 944.47 ms]
Found 23 outliers among 100 measurements (23.00%)
  20 (20.00%) low severe
  3 (3.00%) low mild

Benchmarking distinct query with 1 partitions and 10000000 samples per partition with limit 10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 105.1s, or reduce sample count to 10.
distinct query with 1 partitions and 10000000 samples per partition with limit 10
                        time:   [1.0396 s 1.0424 s 1.0454 s]
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

#### new
Benchmarking distinct query with 100 partitions and 100000 samples per partition with limit 10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 8.9s, or reduce sample count to 50.
distinct query with 100 partitions and 100000 samples per partition with limit 10
                        time:   [85.114 ms 87.070 ms 88.822 ms]
                        **change: [-44.085% -42.690% -41.525%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 20 outliers among 100 measurements (20.00%)
  16 (16.00%) low severe
  4 (4.00%) low mild

Benchmarking distinct query with 10 partitions and 1000000 samples per partition with limit 10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.2s, or reduce sample count to 40.
distinct query with 10 partitions and 1000000 samples per partition with limit 10
                        time:   [104.73 ms 106.50 ms 108.28 ms]
                        **change: [-88.813% -88.594% -88.348%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild

Benchmarking distinct query with 1 partitions and 10000000 samples per partition with limit 10: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 97.1s, or reduce sample count to 10.
distinct query with 1 partitions and 10000000 samples per partition with limit 10
                        time:   [952.15 ms 954.94 ms 957.80 ms]
                        **change: [-8.7663% -8.3904% -8.0083%] (p = 0.00 < 0.05)**
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

File: datafusion/physical-plan/src/aggregates/row_hash.rs [hunk 1]
```diff
@@ -266,6 +266,12 @@ pub(crate) struct GroupedHashAggregateStream {
 
     /// The spill state object
     spill_state: SpillState,
+
+    /// Optional soft limit on the number of `group_values` in a batch
+    /// If the number of `group_values` in a single batch exceeds this value,
+    /// the `GroupedHashAggregateStream` operation immediately switches to
+    /// output mode and emits all groups.
+    group_values_soft_limit: Option<usize>,
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/row_hash.rs:274]: 👍

**Review by alamb — COMMENTED** [review]
This is looking very neat @msirek

File: datafusion/physical-plan/src/aggregates/row_hash.rs (hunk 1 above)
Comment (msirek) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:274]: Thanks. I believe this is ready for a look. Please let me know if you think I should break it down into smaller PRs.

**Comment by alamb** [comment]
Thank you @msirek  -- I hope to review it more carefully later today or tomorrow

File: datafusion/core/benches/data_utils/mod.rs [hunk 2]
```diff
@@ -156,3 +161,83 @@ pub fn create_record_batches(
         })
         .collect::<Vec<_>>()
 }
+
+/// Create time series data with `partition_cnt` partitions and `sample_cnt` rows per partition
```
Comment (alamb) [inline datafusion/core/benches/data_utils/mod.rs:165]: Thank you for adding the comments here

File: datafusion/core/src/physical_optimizer/optimizer.rs [hunk 3]
```diff
@@ -79,6 +80,8 @@ impl PhysicalOptimizer {
             // repartitioning and local sorting steps to meet distribution and ordering requirements.
             // Therefore, it should run before EnforceDistribution and EnforceSorting.
             Arc::new(JoinSelection::new()),
+            // The LimitedDistinctAggregation rule should be applied before the EnforceDistribution rule
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/optimizer.rs:83]: I think adding the rationale for this limitation would be helpful. Your PR description I think explains it pretty well:

```suggestion
            // The LimitedDistinctAggregation rule should be applied before the EnforceDistribution rule
            // As that rule may inject other operations in between the different AggregateExecs.
            // Applying the rule early means only directly-connected AggregateExecs must be examined.
            
```

File: datafusion/core/src/physical_optimizer/combine_partial_final_agg.rs [hunk 4]
```diff
@@ -86,7 +86,7 @@ impl PhysicalOptimizerRule for CombinePartialFinalAggregate {
                                             } else {
                                                 AggregateMode::SinglePartitioned
                                             };
-                                        AggregateExec::try_new(
+                                        let combined_agg = AggregateExec::try_new(
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/combine_partial_final_agg.rs:89]: Another way to express the same logic with less indenting is:

```rust
                                        AggregateExec::try_new(
                                            mode,
                                            input_agg_exec.group_by().clone(),
                                            input_agg_exec.aggr_expr().to_vec(),
                                            input_agg_exec.filter_expr().to_vec(),
                                            input_agg_exec.order_by_expr().to_vec(),
                                            input_agg_exec.input().clone(),
                                            input_agg_exec.input_schema().clone(),
                                        )
                                        .map(|combined_agg| {
                                            combined_agg.with_limit(agg_exec.limit())
                                        })
                                        .ok()
                                        .map(Arc::new)
```

File: datafusion/core/src/physical_optimizer/topk_aggregation.rs [hunk 5]
```diff
@@ -118,25 +118,14 @@ impl TopKAggregation {
...
-        let child = transform_down_mut(child.clone(), &mut closure).ok()?;
+        let child = child.clone().transform_down_mut(&mut closure).ok()?;
         let sort = SortExec::new(sort.expr().to_vec(), child)
             .with_fetch(sort.fetch())
             .with_preserve_partitioning(sort.preserve_partitioning());
         Some(Arc::new(sort))
     }
 }
 
-fn transform_down_mut<F>(
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/topk_aggregation.rs:129]: 👍  for moving into the trait

File: datafusion/core/tests/sql/group_by.rs [hunk 6]
```diff
@@ -70,6 +70,45 @@ async fn group_by_date_trunc() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn distinct_group_by_limit() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/tests/sql/group_by.rs:74]: Does this test add additional coverage compared to the tests in `datafusion/sqllogictest/test_files/aggregate.slt`?

File: datafusion/core/tests/sql/group_by.rs [hunk 7]
```diff
@@ -70,6 +70,45 @@ async fn group_by_date_trunc() -> Result<()> {
...
+"#
+    .trim();
+    let actual = format!("{}", pretty_format_batches(&batches)?);
+    assert_eq!(actual, expected);
+
+    Ok(())
+}
+
 #[tokio::test]
 async fn group_by_limit() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/tests/sql/group_by.rs:113]: 🤔  I think we could port the rest of these tests to group_by.slt 🤔 (not in this PR, I just noticed this while reviewing)

File: datafusion/physical-plan/src/aggregates/mod.rs [hunk 8]
```diff
@@ -800,6 +807,39 @@ impl AggregateExec {
...
+        // ensure there is a group by
+        if self.group_by().is_empty() {
+            return false;
+        }
+        // ensure there are no aggregate expressions
+        if !self.aggr_expr().is_empty() {
+            return false;
+        }
+        // ensure there are no filters on aggregate expressions; the above check
+        // may preclude this case
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/mod.rs:825]: I agree I don't think you can have a filter without an aggregate expression

File: datafusion/physical-plan/src/aggregates/mod.rs [hunk 9]
```diff
@@ -800,6 +807,39 @@ impl AggregateExec {
...
+        if self.filter_expr().iter().any(|e| e.is_some()) {
+            return false;
+        }
+        // ensure there are no order by expressions
+        if self.order_by_expr().iter().any(|e| e.is_some()) {
+            return false;
+        }
+        // ensure there is no output ordering; can this rule be relaxed?
```
Comment (alamb) [inline datafusion/physical-plan/src/aggregates/mod.rs:833]: it is probably subsumed by the check on the required input ordering, because the group operator doesn't introduce any new orderings

File: datafusion/physical-plan/src/aggregates/row_hash.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-plan/src/aggregates/row_hash.rs:274]: I can't see any good way to break this down (other than maybe breaking out the benchmarks), so this is fine

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 10]
```diff
@@ -2504,6 +2504,204 @@ NULL 0 0
...
+physical_plan
+ProjectionExec: expr=[MAX(aggregate_test_100.c1)@2 as MAX(aggregate_test_100.c1), c2@0 as c2, c3@1 as c3]
+--GlobalLimitExec: skip=0, fetch=5
+----AggregateExec: mode=Final, gby=[c2@0 as c2, c3@1 as c3], aggr=[MAX(aggregate_test_100.c1)]
+------CoalescePartitionsExec
+--------AggregateExec: mode=Partial, gby=[c2@1 as c2, c3@2 as c3], aggr=[MAX(aggregate_test_100.c1)]
+----------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
+------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c1, c2, c3], has_header=true
+
+# TODO(msirek): Extend checking in LimitedDistinctAggregation equal groupings to ignore the order of columns
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/aggregate.slt:2621]: BTW the `equivalence` module (recently worked on from @ozankabak and @mustafasrepo ) https://github.com/apache/arrow-datafusion/blob/15d8c9bf48a56ae9de34d18becab13fd1942dc4a/datafusion/physical-expr/src/equivalence.rs has logic to perform this type of analysis

File: datafusion/common/src/config.rs [hunk 11]
```diff
@@ -427,6 +427,11 @@ config_namespace! {
 config_namespace! {
     /// Options related to query optimization
     pub struct OptimizerOptions {
+        /// When set to true, the optimizer will push a limit operation into
+        /// grouped aggregations which have no aggregate expressions, as a soft limit,
+        /// emitting groups once the limit is reached, before all rows in the group are read.
+        pub enable_distinct_aggregation_soft_limit: bool, default = true
```
Comment (alamb) [inline datafusion/common/src/config.rs:433]: 💯  for a disable flag

**Review by alamb — APPROVED** [review]
Wow -- thank you for this contribution @msirek  -- it is very nice read. I found it well documented and well tested. 🦾 

I have a suggestion on how to simplify this code in https://github.com/msirek/arrow-datafusion/pull/1 but I think what is in this PR is also correct and thus can be merged

BTW I would love to know anything you can share about your usecase and what you are doing with DataFusion

cc @avantgardnerio  @thinkharderdev  and @Dandandan  as this is somewhat related to the high cardinality tracing_id usecase

**Review by avantgardnerio — APPROVED** [review]
nice!

**Comment by msirek** [comment]
> I have a suggestion on how to simplify this code in https://github.com/msirek/arrow-datafusion/pull/1

Thanks! I've included those commits in this PR.

> BTW I would love to know anything you can share about your usecase and what you are doing with DataFusion

I'm not really a DataFusion user. I just have a personal interest in query optimization and DataFusion looks pretty neat.

File: datafusion/core/src/physical_optimizer/optimizer.rs (hunk 3 above)
Comment (msirek) [reply datafusion/core/src/physical_optimizer/optimizer.rs]: > I think adding the rationale for this limitation would be helpful. Your PR description I think explains it pretty well:

Made the suggested change.

File: datafusion/core/src/physical_optimizer/combine_partial_final_agg.rs (hunk 4 above)
Comment (msirek) [reply datafusion/core/src/physical_optimizer/combine_partial_final_agg.rs]: > Another way to express the same logic with less indenting is:
> 
> ```rust
>                                         AggregateExec::try_new(
>                                             mode,
>                                             input_agg_exec.group_by().clone(),
>                                             input_agg_exec.aggr_expr().to_vec(),
>                                             input_agg_exec.filter_expr().to_vec(),
>                                             input_agg_exec.order_by_expr().to_vec(),
>                                             input_agg_exec.input().clone(),
>                                             input_agg_exec.input_schema().clone(),
>                                         )
>                                         .map(|combined_agg| {
>                                             combined_agg.with_limit(agg_exec.limit())
>                                         })
>                                         .ok()
>                                         .map(Arc::new)
> ```

Applied this change from your example PR, thanks!

File: datafusion/core/tests/sql/group_by.rs (hunk 6 above)
Comment (msirek) [reply datafusion/core/tests/sql/group_by.rs]: Not really, except that it uses `mode=Single`.  Removed the test.

File: datafusion/physical-plan/src/aggregates/mod.rs (hunk 9 above)
Comment (msirek) [reply datafusion/physical-plan/src/aggregates/mod.rs:833]: > it is probably subsumed by the check on the required input ordering

OK. Even if there is an ORDER BY in a nested expression (e.g derived table), there's no requirement that the rows are presented in that order unless there's a top-level ORDER BY. 

Is it OK to keep this check in until we have a known test case where there is an output ordering but no required input ordering?  Just want to avoid incorrect results in the event there is some edge case not considered.

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 10 above)
Comment (msirek) [reply datafusion/sqllogictest/test_files/aggregate.slt:2621]: > BTW the `equivalence` module (recently worked on from @ozankabak and @mustafasrepo ) https://github.com/apache/arrow-datafusion/blob/15d8c9bf48a56ae9de34d18becab13fd1942dc4a/datafusion/physical-expr/src/equivalence.rs has logic to perform this type of analysis

Thanks, I will take a look.

**Comment by msirek** [comment]
Thanks for the reviews!

**Comment by alamb** [comment]
Screenshot 2023-11-08 at 5 21 38 PM
🤔  this branch has some conflicts that need to be fixed

Perhaps #8004

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 10 above)
Comment (msirek) [reply datafusion/sqllogictest/test_files/aggregate.slt:2640]: It would probably take a bit more work and testing to support this since `PhysicalGroupBy` isn't exactly the same as an equivalence class. I've opened issue #8101 for this.

**Comment by alamb** [comment]
Thanks again @msirek  -- looks great ❤️
